tabular/src/autogluon/tabular/models/vowpalwabbit/vowpalwabbit_model.py

```diff
@@ -0,0 +1,288 @@
+        # while initialising the model.
+        # use_different_namespace: Used for generating different namespace for features
+        # passes: Used as epoch
+
+        ag_params = self._get_ag_params()
+        params = self._get_model_params()
+        namespace_separator = ''
+        if self.use_different_namespace_key in params:
+            namespace_separator = '|'
+            params.pop(self.use_different_namespace_key)
+        params['loss_function'] = params.get('loss_function', self._get_default_loss_function())
+        print(f'Hyperparameters: {params}')
+        epochs = params['passes']
+        params.pop('passes')
+
+        # Make sure to call preprocess on X near the start of `_fit`.
+        # This is necessary because the data is converted via preprocess during predict, and needs to be in the same format as during fit.
+        X_series = self.preprocess(X, is_train=True, namespace_separator=namespace_separator)
+
+        self._validate_loss_function(hyperparams=params)
+
+        # VW expects label from 1 to N for Binary and Multiclass classification problems
+        # AutoGluon does label encoding from 0 to N-1, hence we increment the value of y by 1
+        if self.problem_type != REGRESSION:
+            y = y.apply(lambda row: row + 1)
+        y = y.astype(str) + ' '
+
+        # Concatenate y and X to get the training data in VW format
+        final_training_data = y + X_series
+        final_training_data = final_training_data.tolist()
+
+        # Initialize the model
+        if self.problem_type in PROBLEM_TYPES_CLASSIFICATION:
+            # Ref: https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Predicting-probabilities#multi-class---oaa
+            self.labels = list(set(y.tolist()))
+            self.oaa: int = len(self.labels)
+            print(f'Setting oaa to {self.oaa}')
+            self.model = pyvw.vw(
+                **params,
+                oaa=self.oaa,
+                cache_file='train.cache',
+                holdout_off=True,
+                probabilities=True,  # Only to be used for classification
+                quiet=True,
+            )
+        elif self.problem_type in PROBLEM_TYPES_REGRESSION:
+            self.model = pyvw.vw(
+                **params,
+                cache_file='train.cache',
+                holdout_off=True,
+                quiet=True,
+            )
+        # Train the model
+        for epoch in range(1, epochs + 1):
+            # TODO: Add Early Stopping support
+            self._train_single_epoch(training_data=final_training_data)
+
+        print('Exiting the `_fit` method')
+
+    def _train_single_epoch(self, training_data):
+        np.random.seed(42)
+        row_order = np.arange(0, len(training_data))
+        row_order = np.random.permutation(row_order)
+        for row_i in row_order:
+            row = training_data[row_i]
+            self.model.learn(row)
```

Comment:
given this is a fixed order every time since we are setting the seed each epoch, wouldn't we want to simply reorder the rows once outside of this context and pass that in directly without shuffling each epoch?

Reply:
Seed is being set so that the results are reproducible. For every epoch, order of the rows can be different. Setting seed would ensure that the order would be reproducible for different epochs.

Reply:
From the code you provide here, how would the rows be different if you are resetting the seed to 42 before every call to np.random.permutation? Every epoch will have the same rows.